Sources/Growth/P3ATimedStorage.swift

```diff
@@ -0,0 +1,157 @@
+      purge()
+    }
+  }
+  
+  /// The storage container persisting data for this p3a question
+  private let storage: Preferences.Option<Data>
+  
+  /// Creates a storage container to hold onto P3A data for up to a given number of days
+  public init(name: String, lifetimeInDays: Int) {
+    self.name = name
+    self.lifetimeInDays = lifetimeInDays
+    self.storage = .init(key: "p3a.event-storage.\(name)", default: .init())
+    
+    if let data = try? JSONDecoder().decode([Record].self, from: storage.value) {
+      records = data
+    }
+    purge()
+  }
+  
+  fileprivate var cutoffTime: Date {
+    Calendar.current.date(
+      byAdding: .day,
+      value: -lifetimeInDays,
+      to: Calendar.current.startOfDay(for: date())
+    ) ?? Date()
+  }
+  
+  /// Purges any records that are outside of the lifetime
+  private mutating func purge() {
+    records.removeAll(where: { [cutoffTime] in $0.date <= cutoffTime })
+  }
+  
+  /// Resets the storage container
+  public mutating func reset() {
+    storage.reset()
+    records = []
+  }
+  
+  /// Adds a record to the list
+  public mutating func append(value: Value) {
+    defer {
+      save()
+    }
+    records.append(.init(id: .init(), date: Calendar.current.startOfDay(for: date()), value: value))
+  }
+  
+  fileprivate mutating func save() {
+    purge()
+    if let data = try? JSONEncoder().encode(records) {
+      storage.value = data
+    }
+  }
+}
+
+extension P3ATimedStorage.Record: Equatable where Value: Equatable {}
+extension P3ATimedStorage.Record: Hashable where Value: Hashable {}
+
+extension P3ATimedStorage where Value: Comparable {
+  /// Gets the smallest value out of all the records stored
+  public var minimumValue: Value? {
+    records.map(\.value).min()
+  }
+  /// Gets the largest value out of all the records stored
+  public var maximumValue: Value? {
+    records.map(\.value).max()
+  }
+  /// Appends a value if the value given is larger than any value recorded today
+  public mutating func replaceTodaysRecordsIfLargest(value: Value) {
+    let today = Calendar.current.startOfDay(for: date())
+    let todaysRecords = records.filter({ Calendar.current.startOfDay(for: $0.date) == today })
+    guard let maxValue = todaysRecords.map(\.value).max() else {
+      append(value: value)
+      return
+    }
+    if value > maxValue {
+      records.removeAll(where: { record in todaysRecords.contains(where: { $0.id == record.id })})
+      append(value: value)
+    }
+  }
+}
+
+extension P3ATimedStorage where Value: AdditiveArithmetic & Comparable {
+  /// Returns the sum of all the records values or 0 if no values are recorded
+  public var combinedValue: Value {
+    let values = records.map(\.value)
+    if values.isEmpty { return .zero }
+    return max(.zero, values.reduce(.zero, { $0 + $1 }))
+  }
+  /// Groups and combines the values of each day's values recorded then returns the largest value in that set
+  public var maximumDaysCombinedValue: Value {
```

Comment:
maybe we can add couple of unit tests to the p3a api you added if not too hard to add

Reply:
Added a unit test for `maximumDaysCombinedValue`, the rest were already unit tested